dex/src/main/scala/com/wavesplatform/dex/db/OrderDB.scala

```diff
@@ -1,86 +1,101 @@
-            rw.get(DbKeys.finalizedPair(sender, oi.assetPair, newPairSeqNr - settings.maxOrders))
-              .map(DbKeys.order)
-              .foreach(x => rw.delete(x))
-
-          rw.put(orderInfoKey, Some(oi))
-        }
+      for {
+        hasKey <- levelDb.has(orderInfoKey)
+        _ <-
+          F.whenA(!hasKey) {
+            levelDb.readWrite { rw =>
+              val newCommonSeqNr = rw.inc(DbKeys.finalizedCommonSeqNr(sender))
+              rw.put(DbKeys.finalizedCommon(sender, newCommonSeqNr), Some(id))
+
+              val newPairSeqNr = rw.inc(DbKeys.finalizedPairSeqNr(sender, oi.assetPair))
+              rw.put(DbKeys.finalizedPair(sender, oi.assetPair, newPairSeqNr), Some(id))
+              if (newPairSeqNr > settings.maxOrders) // Indexes start with 1, so if maxOrders=100 and newPairSeqNr=101, we delete 1 (the first)
+                rw.get(DbKeys.finalizedPair(sender, oi.assetPair, newPairSeqNr - settings.maxOrders))
+                  .map(DbKeys.order)
+                  .foreach(x => rw.delete(x))
+
+              rw.put(orderInfoKey, Some(oi))
+            }
+          }
+      } yield ()
     }
 
-    override def getFinalizedOrders(owner: Address, maybePair: Option[AssetPair]): Seq[(Order.Id, OrderInfo[OrderStatus])] =
-      db.readOnly { ro =>
-        val (seqNr, key) = maybePair match {
-          case Some(p) =>
-            (ro.get(DbKeys.finalizedPairSeqNr(owner, p)), DbKeys.finalizedPair(owner, p, _: Int))
-          case None =>
-            (ro.get(DbKeys.finalizedCommonSeqNr(owner)), DbKeys.finalizedCommon(owner, _: Int))
+    override def getFinalizedOrders(owner: Address, maybePair: Option[AssetPair]): F[Seq[(Order.Id, OrderInfo[OrderStatus])]] =
+      for {
+        (seqNr, key) <- levelDb.readOnly { ro =>
+          maybePair match {
+            case Some(p) =>
+              (ro.get(DbKeys.finalizedPairSeqNr(owner, p)), DbKeys.finalizedPair(owner, p, _: Int))
+            case None =>
+              (ro.get(DbKeys.finalizedCommonSeqNr(owner)), DbKeys.finalizedCommon(owner, _: Int))
+          }
         }
-
-        (for {
-          offset <- 0 until math.min(seqNr, settings.maxOrders)
-          id <- db.get(key(seqNr - offset))
-          oi <- db.get(DbKeys.orderInfo(id))
-        } yield id -> oi).sorted
-      }
-
-    override def getOrderInfo(id: Id): Option[FinalOrderInfo] = db.readOnly(_.get(DbKeys.orderInfo(id)))
-
-    override def transactionsByOrder(orderId: Id): Seq[ExchangeTransaction] = db.readOnly { ro =>
+        offsets = 0 until math.min(seqNr, settings.maxOrders)
+        ids <- offsets
+          .toList
+          .traverse(offset => levelDb.get(key(seqNr - offset)))
+          .map(_.flatten)
+        results <- ids
+          .traverse(id => levelDb.get(DbKeys.orderInfo(id)).map(maybeOrderInfo => maybeOrderInfo.map(x => id -> x)))
+          .map(_.flatten)
```

Comment:
This could be done within one LevelDB's snapshot. I mean, could you try to use `ro.get` instead of `levelDb.get` to get all data and sort the result outside the `levelDb.readOnly`?

Reply:
yea makes sense... resolved

Reply:
> … and sort the result outside the levelDb.readOnly 

;)